### ISSUE TITLE
transporting along identity paths

### ISSUE THREAD

**Issue by fpvandoorn** [leanprover/lean#325]
The term `(idpath a) ▹ b'` is definitionally equal to `b'`. However, if I replace `b'` by `(p ▹ b)`, Lean doesn't recognize anymore that these are definitionally equal:

```
import hott.path
open path
variables {A : Type} {B : A → Type} {a a' : A} {b : B a} {b' : B a'} (p : a ≈ a')
check (λ(r : idp ▹ b' ≈ b'), r) idp -- this works
check (λ(r : idp ▹ (p ▹ b) ≈ (p ▹ b)), r) idp --error
check (λ(r : transport B (idpath a') (transport B p b) ≈ (transport B p b)), r) (idpath (transport B p b)) --giving some implicit arguments doesn't solve error
print "\n  some more tests:"
check (λ(b' : B a') (r : (idpath a) ▹ b' ≈ b'), r) b' (idpath b')
check (λ(b' : B a') (r : (idpath a) ▹ b' ≈ b'), r) (p ▹ b) (idpath (p ▹ b)) --error
print ""
check (λ(b' : B a'), ((λ(r : idp ▹ b' ≈ b'), r) idp)) b'
check (λ(b' : B a'), ((λ(r : idp ▹ b' ≈ b'), r) idp)) (p ▹ b) -- not an error!
```

Note: in the last line, I end up with the same term (up to beta conversion), but that does typecheck.

Is this a bug?

EDIT: Is there a better way to ask Lean "does `a` has type `A`?" than writing `check (λ x : A, x) a`?

**Comment by leodemoura** [comment]
I will investigate.

> EDIT: Is there a better way to ask Lean "does a has type A?" than writing check (λ x : A, x) a?

We can also use

```
check typeof a : A
```

It is not much better :-( 
`typeof` is a notation defined at general_notation.lean

```
definition is_typeof (A : Type) (a : A) : A := a
notation `typeof` t `:` T  := is_typeof T t
```

**Comment by fpvandoorn** [comment]
Ok, great! 
I like `check typeof a : A` much better. The syntax looks a bit as if `check typeof` is an actual command (modulo highlighting) :-)

**Comment by leodemoura** [comment]
I finally found out what is going on.
This is not a bug. The issue is that `transport` is not marked as `[reducible]`.
So, some definitional equalities are ignored by the elaborator when it is trying to solve higher-order unification problems. All examples work if we add the command:

```
reducible transport
```

I think any "recursor-like" definition such as `transport` should be marked as `[reducible]`. This is what I'm doing with all automatically generated recursors on top of `rec` (e.g., `cases_on`, `rec_on`, `induction_on`, etc). Similarly, any "projection-like" definitions should be also marked as `[reducible]`. I'm also doing that in the `structure` command. Otherwise, we get counter-intuitive behavior like the one you describe.

**Comment by leodemoura** [comment]
Remark: fixed some typos in the comment above.

**Comment by fpvandoorn** [comment]
Ah. Good that you could solve this. Yes, it kind-of makes sense that if the elaborator cannot unfold transport, it will fail. I'll make transport reducible.

So what is the heuristic for which theorems should be `reducible` (and why)? I don't really have a good sense what makes "recursor-like" and "projection-like" definitions different to other definitions, which causes problems. 
Or maybe an easier question: should the following definitions be reducible? (I think both shouldn't be, but I can't tell why they are less likely to cause problems like transport does).

```
-- special case of transport, no higher order unification involved
definition ap ⦃A B : Type⦄ (f : A → B) {x y:A} (p : x ≈ y) : f x ≈ f y :=
path.rec_on p idp

-- properties of transport, like the following:
definition moveR_transport_p {A : Type} (P : A → Type) {x y : A} (p : x ≈ y) (u : P x) (v : P y) :
  u ≈ p⁻¹ ▹ v → p ▹ u ≈ v :=
rec_on p (take v, id) v
```

**Comment by leodemoura** [comment]
> So what is the heuristic for which theorems should be reducible (and why)? 
> I don't really have a good sense what makes "recursor-like" and "projection-like"
> definitions different to other definitions, which causes problems. 

Both of them (recursor and projection) are auxiliary definitions. That is, we use them to build the "interesting" declarations for a datatype `T`. If they are not reducible, then we would have two options when defining a declaration `D` (using auxiliary definitions):

1- We would have to define `D` in the same file where `T` is defined (the auxiliary definitions are transparent there).

2- We would not be able to rely on "computation" in higher-order unification constraints. If `D` is defined in a different file, then the auxiliary definitions would not be considered reducible.

This seems odd. Let's consider the auxiliary recursors: `T.rec_on`, `T.cases_on`, `T.induction_on`, etc.
We can view them as convenience for helping us use `T.rec`. If we don't mark them reducible, then we would get completely different behavior whether we use `T.rec` or one of the auxiliary recursors.

Short version: if we are declaring an auxiliary definition that will be used to define a lot of things in different files, then we should probably mark it as reducible.

**Comment by leodemoura** [comment]
> Or maybe an easier question: should the following definitions be reducible? (I think both shouldn't be, but I can't tell why they are less likely to cause problems like transport does).

I think both of them should be. At least `ap`, since it is used to define many other things, right?

**Comment by leodemoura** [comment]
I will close this issue, since the problem was solved using `[reducible]`. We can reopen the issue if we still have problems.
